### ISSUE TITLE
Firewall-cmd is extremely slow when populating large ipsets

### ISSUE THREAD

**Issue by Jiivee1** [firewalld/firewalld#881]
**What happened**: It takes about 78 minutes to add 19000 entries to existing, but empty ipset.

**What you expected to happen**: Adding 19000 entries should take less than minute.

**How to reproduce it (as minimally and precisely as possible)**:

Prerequisite: have 19000 networks in ./blacklist_nets_sorted file,
firewall-cmd --permanent --new-ipset=deny_public --type=hash:net --option=family=inet --option=hashsize=16384 --option=maxelem=20000
firewall-cmd --permanent --ipset=deny_public --add-entries-from-file=./blacklist_nets_sorted

**Anything else we need to know?**: Testing with smaller number of added networks to ipset shows that when number entries doubles, time to process them increases about four fold. Firewall-cmd spends almost all time in userland. *) Tested with similar results with iptables and nftables backends.

**Environment**:
- Firewalld Version (if Fedora based `dnf info firewalld` or commit hash if developing from git `git log -n1 --format=format:"%H"`): 1.0.1
- Firewalld Backend (`cat /etc/firewalld/firewalld.conf | grep FirewallBackend`): iptables *)
- OS (e.g: `cat /etc/os-release`): NAME="openSUSE Tumbleweed"
VERSION="20211019"
ID="opensuse-tumbleweed"
ID_LIKE="opensuse suse"
VERSION_ID="20211019"
PRETTY_NAME="openSUSE Tumbleweed"
ANSI_COLOR="0;32"
CPE_NAME="cpe:/o:opensuse:tumbleweed:20211019"
BUG_REPORT_URL="https://bugs.opensuse.org"
HOME_URL="https://www.opensuse.org/"
DOCUMENTATION_URL="https://en.opensuse.org/Portal:Tumbleweed"
LOGO="distributor-logo-Tumbleweed"
- Others: See also: https://forums.opensuse.org/showthread.php/561722-Firewall-cmd-unable-to-populate-large-ipset

**Comment by Jiivee1** [comment]
I researched this a bit further down to isolate where slowness is caused. In /usr/bin/firewall-cmd, from line 1285 onward:
```
            for filename in a.remove_entries_from_file:
                try:
                    entries = cmd.get_ipset_entries_from_file(filename)
                except IOError as msg:
                    message = "Failed to read file '%s': %s" % (filename, msg)
                    if len(a.remove_entries_from_file) > 1:
                        cmd.print_warning(message)
                    else:
                        cmd.print_and_exit(message)
                else:
                    old_entries = settings.getEntries()
                    entries_set = set()
                    for entry in old_entries:
                        entries_set.add(entry)
                    for entry in entries:
                        if entry in entries_set:
                            old_entries.remove(entry)
                            entries_set.discard(entry)
                            changed = True
                        else:
                            cmd.print_if_verbose("Warning: NOT_ENABLED: %s" % entry)
                    if changed:
                        settings.setEntries(old_entries)
```

When adding large amount of entries to ipset, time is spent in settings.setEntries call on last line of above snippet.  setEntries is defined in /usr/lib/python3.8/site-packages/firewall/client.py in class FirewallClientIPSetSettings around line 1515 as:
```
    @handle_exceptions
    def setEntries(self, entries):
        if "timeout" in self.settings[4] and \
           self.settings[4]["timeout"] != "0":
            raise FirewallError(errors.IPSET_WITH_TIMEOUT)
        _entries = set()
        for _entry in dbus_to_python(entries, list):
            check_entry_overlaps_existing(_entry, _entries)
            _entries.add(normalize_ipset_entry(_entry))
        self.settings[5] = list(_entries)
    @handle_exceptions
```

Further check_entry_overlaps_existing is defined in /usr/lib/python3.8/site-packages/firewall/core/ipset.py onlines 306-317 as:
```
def check_entry_overlaps_existing(entry, entries):
    """ Check if entry overlaps any entry in the list of entries """
    # Only check simple types
    if len(entry.split(",")) > 1:
        return

    for itr in entries:
        try:
            if ipaddress.ip_network(itr, strict=False).overlaps(ipaddress.ip_network(entry, strict=False)):
                raise FirewallError(errors.INVALID_ENTRY, "Entry '{}' overlaps with existing entry '{}'".format(itr, entry))
        except ValueError:
            pass
```

I don't know whether check_entry_overlaps_existing is causing the slowness or not. As I am not familiar with Python, I don't know how to use a local copy instead of those system libs for example to add logging to see what's going on.

In OpenSuse Tumbleweed adding over 200000 routes to ipset without aggregating them in advance used to take about three minutes. On late August 2021 things changed and adding such a large number of entries became impossible (it takes more than a week!). Nowadays it takes about 77 minutes to add ca. 19000 IPv4 routes to ipset and about 12 hours to add ca. 49000 IPv6 routes to ipset.

**Comment by cryptoluks** [comment]
> https://github.com/firewalld/firewalld/issues/881#issuecomment-967122743

Thanks for pinpointing the issue.

It seems, that `check_entry_overlaps_existing` introduced the behavior we observe. Commenting out the usage of this function in `client.py` restores the old behavior. It was added with https://github.com/firewalld/firewalld/issues/881#issuecomment-967122743.

So the whole check seems to be very slow, as the function is called for each entry and each entry will trigger the `for` loop, which iterates every time through the previously imported entries to check if any overlap. This may be the reason, why you [observed](https://forums.opensuse.org/showthread.php/561722-Firewall-cmd-unable-to-populate-large-ipset) the exponential growth in time like behavior.

So we might need a better solution for checking for overlapping IP addresses or simply revert the commit for the time?

What happens if overlapping addresses are imported?

Maybe @erig0 can have a look into this? 😄

**Comment by erig0** [comment]
I'll look at it when I can, but I'm pretty backlogged. Feel free to take a look if you have time. Maybe the overlap detection so it's not so costly.

**Comment by cryptoluks** [comment]
Thanks for your quick response.

I made some rudimentary testing with `IntervalTree`, as this should be more performant. Inspired by [this](https://stackoverflow.com/a/36055588) it could look something like this:

```python
from intervaltree import Interval, IntervalTree

entries_tree = IntervalTree(Interval(ipaddress.ip_network(net).network_address, ipaddress.ip_network(net).broadcast_address + 1) for net in entries)

entries_tree.overlaps(ipaddress.ip_network(entry, strict=False).network_address, ipaddress.ip_network(entry, strict=False).broadcast_address)
```

As already mentioned [here](https://github.com/chaimleib/intervaltree/issues/59#issuecomment-638523196) in our context, `IntervalTree` does not support null intervals (single ip), and therefore we would have to add 1 to the end of the interval as above. I am not entirely sure if works without side effects.

Nevertheless, the root issue still persists. Even with `IntervalTree` it is very slow if we process every imported net on each iteration.

I have two ideas:

1. Make the overlap detection optional if overlapping nets do not do any harm (until a better solution is available, as many users seem to have issues with this at the moment).
2. Do not process the whole list of nets on each iteration, instead, read the whole to-be-imported ipset right away and throw an error if nets do either overlap *within* the ipset or *between* the ipset and previously imported nets. This should be much faster.

What do you think @erig0?

**Comment by erig0** [comment]
Hi all, please give #893 a try. It's a significant improvement, but still not fantastic.

**Comment by erig0** [comment]
No feedback. I'm merging #893 which will close this issue. If it proves insufficient down the road then we can revisit by filing a new issue.

**Comment by erig0** [comment]
> No feedback. I'm merging #893 which will close this issue. If it proves insufficient down the road then we can revisit by filing a new issue.

Note that this has a follow up, #907.

**Comment by cryptoluks** [comment]
Incredible @erig0! I ran a quick test with the improved `ipset.py` of your [fork](https://github.com/erig0/firewalld/blob/9df2d95321f4454ddba18cfe6c2ba30dbeb018bd/src/firewall/core/ipset.py) and it is now again a matter of seconds to create larger ipsets!

Also, importing overlapping ipsets (within a file or within two files) triggers still the overlapping exception.

Didn't expect that sorting the networks will improve `.overlaps`-performance that much.

Well done, thank you 👍

**Comment by erig0** [comment]
Thanks for testing and reviewing!

**Comment by erig0** [comment]
> Didn't expect that sorting the networks will improve `.overlaps`-performance that much.

Take a look at the comment in the commit (#907) for an explanation of how it works.
